### PR TITLE
[SPARK-31768][ML] add getMetrics in Evaluators

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
add getMetrics in Evaluators to get the corresponding Metrics instance, so users can use it to get any of the metrics scores. For example:

```
    val trainer = new LinearRegression
    val model = trainer.fit(dataset)
    val predictions = model.transform(dataset)

    val evaluator = new RegressionEvaluator()

    val metrics = evaluator.getMetrics(predictions)
    val rmse = metrics.rootMeanSquaredError
    val r2 = metrics.r2
    val mae = metrics.meanAbsoluteError
    val variance = metrics.explainedVariance
```

### Why are the changes needed?
Currently, Evaluator.evaluate only access to one metrics, but most users may need to get multiple metrics. This PR adds getMetrics in all the Evaluators, so users can use it to get an instance of the corresponding Metrics to get any of the metrics they want. 

### Does this PR introduce _any_ user-facing change?
Yes. Add getMetrics in Evaluators.
For example:
```
  /**
   * Get a RegressionMetrics, which can be used to get any of the regression
   * metrics such as rootMeanSquaredError, meanSquaredError, etc.
   *
   * @param dataset a dataset that contains labels/observations and predictions.
   * @return RegressionMetrics
   */
  @Since("3.1.0")
  def getMetrics(dataset: Dataset[_]): RegressionMetrics
```

### How was this patch tested?
Add new unit tests

### CODE REVIEW

**Review by srowen — APPROVED** [review]
Looks OK pending tests

**Comment by zhengruifeng** [comment]
I do not object this PR. Is there any way to recompute another metric in evaluator without involving mllib.metrics?

**Comment by huaxingao** [comment]
> Is there any way to recompute another metric in evaluator without involving mllib.metrics?

It might be better to create a ml wrapper for mllib.metrics instead of using it directly, something like the approach in ```LogisticRegressionSummary```. This wrapper can be used both by evaluator and summary.

**Comment by srowen** [comment]
Merged to master

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringMetrics.scala [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+
+  def getDistanceMeasure: String = distanceMeasure
+
+  def setDistanceMeasure(value: String) : Unit = distanceMeasure = value
+
+  /**
+   * Returns the silhouette score
+   */
+  @Since("3.1.0")
+  lazy val silhouette: Double = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringMetrics.scala:46]: This method will trigger I/O. By convention you should have `()` in the defintion.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringMetrics.scala [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+ * Metrics for clustering, which expects two input columns: prediction and label.
+ */
+@Since("3.1.0")
+class ClusteringMetrics private[spark](dataset: Dataset[_]) {
+
+  private var distanceMeasure: String = "squaredEuclidean"
+
+  def getDistanceMeasure: String = distanceMeasure
+
+  def setDistanceMeasure(value: String) : Unit = distanceMeasure = value
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringMetrics.scala:40]: Should set return `self.type`? When we have more setters it is easier to chain.

**Pull request closed without merging**